### ISSUE TITLE
Large (~850ms) stalls on 5000 tables tests

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#21646]
The following stalls in the schema merge path were seen in enterprise release in the 5000 tables test.
Examining the code shows that master (6.3.0-dev) is still susceptible to the same reactor stalls.

```
Processed 4 stalls lasting a total of 3397 milliseconds.
min=848 avg=849.2 median=849 p95=850 p99=None p999=None max=850

This graph is printed in bottom-up order, where callees are printed first.
Use --direction=top-down to print callers first.

[level#index/out_of pct%] below denotes:
  level  - nesting level in the graph
  index  - index of node among to its siblings
  out_of - number of siblings
  pct    - percentage of total stall time of this call relative to its siblings

++[0#1/1 100%] addr=0x4c167cf total=3397 count=4 avg=849:
|              seastar::backtrace<seastar::backtrace_buffer::append_backtrace_oneline()::{lambda(seastar::frame)#1}> at ./build/release/seastar.lto/./seastar/include/seastar/util/backtrace.hh:68
++           - addr=0x4c0f575:
|              seastar::backtrace_buffer::append_backtrace_oneline at ./build/release/seastar.lto/./seastar/src/core/reactor.cc:839
|              (inlined by) seastar::print_with_backtrace at ./build/release/seastar.lto/./seastar/src/core/reactor.cc:858
++           - addr=0x4c0eb97:
|              seastar::internal::cpu_stall_detector::generate_trace at ./build/release/seastar.lto/./seastar/src/core/reactor.cc:1482
++           - addr=0x292eb78:
|              seastar::internal::cpu_stall_detector::maybe_report at ./build/release/seastar.lto/./seastar/src/core/reactor.cc:1219
|              (inlined by) seastar::internal::cpu_stall_detector::on_signal at ./build/release/seastar.lto/./seastar/src/core/reactor.cc:1239
++           - addr=0x3dbaf: ?? ??:0
  ++[1#1/4 25%] addr=0x188b060 total=850 count=1 avg=850:
  |             bytes_ostream::write at ././bytes_ostream.hh:249
  |             (inlined by) bytes_ostream::write at ././bytes_ostream.hh:263
  |             (inlined by) ser::serialize_integral<unsigned int, bytes_ostream> at ././serializer.hh:203
  |             (inlined by) ser::integral_serializer<unsigned int>::write<bytes_ostream> at ././serializer.hh:217
  |             (inlined by) ser::serialize<unsigned int, bytes_ostream> at ././serializer.hh:254
  |             (inlined by) ser::column__c__variant<bytes_ostream>::start_variant_live_cell at ./build/release/gen/idl/mutation.dist.impl.hh:4527
  |             (inlined by) (anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}::operator() at ./mutation/mutation_partition_serializer.cc:111
  |             (inlined by) row::maybe_invoke_with_hash<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1} const, cell_and_hash const> at ./mutation/mutation_partition.hh:133
  |             (inlined by) row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}::operator() at ./mutation/mutation_partition.hh:152
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true>::operator() at ././utils/compact-radix-tree.hh:1888
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::visit_slot<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true>&> at ././utils/compact-radix-tree.hh:1560
  ++          - addr=0x1887eb0:
  |             compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u>::visit<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true>&> at ././utils/compact-radix-tree.hh:1364
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::node_base<cell_and_hash, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::direct_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u> >::visit<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true>&, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::direct_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u> > at ././utils/compact-radix-tree.hh:799
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::node_base<cell_and_hash, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::direct_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u> >::visit<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true>&, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::direct_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u> > at ././utils/compact-radix-tree.hh:802
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::node_base<cell_and_hash, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)1, 4u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)2, 8u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::indirect_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)3, 16u>, compact_radix_tree::tree<cell_and_hash, unsigned int>::direct_layout<cell_and_hash, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)6, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)0, 0u, (compact_radix_tree::tree<cell_and_hash, unsigned int>::layout)4, 32u> >::visit<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true>&> at ././utils/compact-radix-tree.hh:807
  ++          - addr=0x188329b:
  |             compact_radix_tree::tree<cell_and_hash, unsigned int>::node_head::visit<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true> > at ././utils/compact-radix-tree.hh:473
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::visit<compact_radix_tree::tree<cell_and_hash, unsigned int>::walking_visitor<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}, true> > at ././utils/compact-radix-tree.hh:1626
  |             (inlined by) compact_radix_tree::tree<cell_and_hash, unsigned int>::walk<row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}>(ser::deletable_row__cells<bytes_ostream>&&) const::{lambda(unsigned int, cell_and_hash const&)#1}> at ././utils/compact-radix-tree.hh:1909
  |             (inlined by) row::for_each_cell<(anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> >(ser::deletable_row__cells<bytes_ostream>&&, row const&, schema const&, column_kind)::{lambda(unsigned int, atomic_cell_or_collection const&)#1}> at ./mutation/mutation_partition.hh:151
  |             (inlined by) (anonymous namespace)::write_row_cells<ser::deletable_row__cells<bytes_ostream> > at ./mutation/mutation_partition_serializer.cc:98
  |             (inlined by) write_row<ser::writer_of_deletable_row<bytes_ostream> > at ./mutation/mutation_partition_serializer.cc:169
  |             (inlined by) mutation_partition_serializer::write_serialized<ser::writer_of_mutation_partition<bytes_ostream> > at ./mutation/mutation_partition_serializer.cc:181
  ++          - addr=0x16d9e60:
  |             mutation_partition_serializer::write at ./mutation/mutation_partition_serializer.cc:216
  |             (inlined by) canonical_mutation::canonical_mutation(mutation const&)::$_0::operator()<ser::writer_of_mutation_partition<bytes_ostream> > at ./mutation/canonical_mutation.cc:32
  |             (inlined by) ser::after_canonical_mutation__mapping<bytes_ostream>::partition<canonical_mutation::canonical_mutation(mutation const&)::$_0> at ./build/release/gen/idl/mutation.dist.impl.hh:6947
  |             (inlined by) canonical_mutation::canonical_mutation at ./mutation/canonical_mutation.cc:31
  ++          - addr=0x1e64902:
  |             schema_mutations::columns_canonical_mutation at ././schema_mutations.hh:90
  |             (inlined by) ser::serializer<schema_mutations>::write<bytes_ostream> at ./build/release/gen/idl/frozen_schema.dist.impl.hh:53
  |             (inlined by) ser::serialize<schema_mutations, bytes_ostream> at ././serializer.hh:254
  |             (inlined by) ser::after_schema__version<bytes_ostream>::write_mutations at ./build/release/gen/idl/frozen_schema.dist.impl.hh:211
  |             (inlined by) frozen_schema::frozen_schema(seastar::lw_shared_ptr<schema const> const&)::$_0::operator() at ./frozen_schema.cc:22
  |             (inlined by) frozen_schema::frozen_schema at ./frozen_schema.cc:17
  ++          - addr=0x2b5124b:
  |             global_schema_ptr::global_schema_ptr(seastar::lw_shared_ptr<schema const> const&)::$_0::operator()(seastar::lw_shared_ptr<schema const> const&) const::{lambda(utils::tagged_uuid<table_schema_version_tag>)#1}::operator() at ./schema/schema_registry.cc:372
  |             (inlined by) std::__invoke_impl<frozen_schema, global_schema_ptr::global_schema_ptr(seastar::lw_shared_ptr<schema const> const&)::$_0::operator()(seastar::lw_shared_ptr<schema const> const&) const::{lambda(utils::tagged_uuid<table_schema_version_tag>)#1}&, utils::tagged_uuid<table_schema_version_tag> > at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
  |             (inlined by) std::__invoke_r<frozen_schema, global_schema_ptr::global_schema_ptr(seastar::lw_shared_ptr<schema const> const&)::$_0::operator()(seastar::lw_shared_ptr<schema const> const&) const::{lambda(utils::tagged_uuid<table_schema_version_tag>)#1}&, utils::tagged_uuid<table_schema_version_tag> > at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:114
  |             (inlined by) std::_Function_handler<frozen_schema (utils::tagged_uuid<table_schema_version_tag>), global_schema_ptr::global_schema_ptr(seastar::lw_shared_ptr<schema const> const&)::$_0::operator()(seastar::lw_shared_ptr<schema const> const&) const::{lambda(utils::tagged_uuid<table_schema_version_tag>)#1}>::_M_invoke at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
  ++          - addr=0x1853707:
  |             std::function<frozen_schema (utils::tagged_uuid<table_schema_version_tag>)>::operator() at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
  |             (inlined by) schema_registry::get_or_load at ./schema/schema_registry.cc:153
  ++          - addr=0x1545337:
  |             global_schema_ptr::global_schema_ptr(seastar::lw_shared_ptr<schema const> const&)::$_0::operator() at ./schema/schema_registry.cc:371
  |             (inlined by) global_schema_ptr::global_schema_ptr at ./schema/schema_registry.cc:377
  | ++[2#1/1 100%] addr=0x259b621 total=2547 count=3 avg=849:
  | |              std::construct_at<global_schema_ptr, seastar::lw_shared_ptr<schema const>&> at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/stl_construct.h:97
  | |              (inlined by) std::allocator_traits<std::allocator<global_schema_ptr> >::construct<global_schema_ptr, seastar::lw_shared_ptr<schema const>&> at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/alloc_traits.h:540
  | |              (inlined by) std::vector<global_schema_ptr, std::allocator<global_schema_ptr> >::emplace_back<seastar::lw_shared_ptr<schema const>&> at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/vector.tcc:117
  | |              (inlined by) db::schema_tables::diff_table_or_view at ./db/schema_tables.cc:1480
  | ++           - addr=0x259491a:
  | |              db::schema_tables::merge_tables_and_views at ./db/schema_tables.cc:1515
  |   ++[3#1/1 100%] addr=0x2578b1a total=3397 count=4 avg=849:
  |   |              db::schema_tables::do_merge_schema at ./db/schema_tables.cc:1361
```

**Comment by ndragazis** [comment]
> Examining the code shows that master (6.3.0-dev) is still susceptible to the same reactor stalls.

Indeed, I managed to reproduce it from OSS. The backtrace is a bit stale since the schema merging code has been moved from `schema_tables.cc` to `schema_applier.cc`.

As in the enterprise release, the stalls come from `db::schema_tables::diff_table_or_view()`. This is a **synchronous** function that loops over a given set of schema mutations and builds schemas for them. When bootstrapping a new node, the new node calls this function as part of merging the schemas that it pulls from the cluster. In our case, this function has to create schemas for 5000 tables and 5000 views (one per index), thus stalling the reactor.

In more detail, here is what happens step-by-step:
1. The SCT runner spins up a Scylla cluster with one node.
2. Due to `pre_create_schema: true` in the test's config yaml, the longevity test creates the schema upfront. The schema consists of 5000 tables and indexes.
3. Due to `add_node_cnt: 5`, the longevity test adds 5 more nodes to the cluster.
4. When joining each new node via Raft, the leader sends its latest snapshot descriptor to the new node and the new node pulls the snapshot from the leader via the migration manager.
5. Each new node applies the snapshot locally by merging the schema changes. The call chain is the following:
    ```
   db::schema_tables::diff_table_or_view
   db::schema_tables::merge_tables_and_views
   db::schema_tables::do_merge_schema
   db::schema_tables::merge_schema
   service::migration_manager::merge_schema_from
   service::group0_state_machine::transfer_snapshot
   service::raft_rpc::apply_snapshot
    ```
6. The snapshot carries mutations for 5000 tables and views. `diff_table_or_view()` builds schemas for all of them. This is a synchronous operation with no yields in-between.

This is not a regression as `diff_table_or_view` was introduced in 2017 (https://github.com/scylladb/scylladb/commit/be12a2bf0a8).

**Comment by bhalevy** [comment]
@ndragazis it should be pretty straightforward to coroutinize `diff_table_or_view` and
add calls to `maybe_yield` in the loops.

If needed, the `create_schema` function passed to can also be made async and yield
if creating a single table / view schema may stall too, but I doubt it.

**Comment by ndragazis** [comment]
> @ndragazis it should be pretty straightforward to coroutinize `diff_table_or_view` and
add calls to `maybe_yield` in the loops.

Yes, I've already tried that. It works.

I'm taking a closer look at the code to make sure I'm not missing anything. The patch https://github.com/scylladb/scylladb/commit/be12a2bf0a8 speaks about "atomic publishes of base and view changes". The comment above `merge_tables_and_views()` reads (emphasis mine):
> Atomically publishes schema changes. In particular, this function ensures
> that when a base schema and a subset of its views are modified together (i.e.,
> upon an alter table or alter type statement), then they are published together
> as well, **without any deferring in-between**.

**Comment by ndragazis** [comment]
The merge lock indeed serializes schema changes on shard 0, but what about read queries happening concurrently with schema changes?

If we turn `diff_table_or_view()` into a coroutine, the new schema versions for a base table and its views may be published (i.e., added to the schema registry) non-atomically. But this can happen anyway if the query is served by another shard, so I'm not sure if it's an issue.

Anyway, I guess if the tests pass, we are good to go.

**Comment by bhalevy** [comment]
> The merge lock indeed serializes schema changes on shard 0, but what about read queries happening concurrently with schema changes?
> 
> If we turn `diff_table_or_view()` into a coroutine, the new schema versions for a base table and its views may be published (i.e., added to the schema registry) non-atomically. But this can happen anyway if the query is served by another shard, so I'm not sure if it's an issue.

It's not a signifucant issue.  Today we may already yield between updating different parts of the overall schema.
If we ever want ti make the new schema published together making the dunctions synchronous is not the right direction in any case, as seen here.
The preparation phase that mat block needs to be able to yield, and only all the different paets of the schema are updated using temporary objects those can be applied synchronously, assuming there is not much cpu consumed in the apply phase and no memory allocations.

> 
> Anyway, I guess if the tests pass, we are good to go.

yup

**Comment by kbr-scylla** [comment]
Please be careful, I don't want to spend another dozens of hours triaging issues like https://github.com/scylladb/scylladb/issues/17786#issuecomment-2059213151 and https://github.com/scylladb/scylladb/issues/18709 caused by "innocent" yields being introduced, e.g. https://github.com/scylladb/scylladb/issues/17786 was caused by https://github.com/scylladb/scylladb/commit/f2ff701489995fdc84ea7ca0addbec2918c35be4 IIRC

>     Anyway, I guess if the tests pass, we are good to go.

Definitely not, this is the wrong approach when talking about concurrency issues. Tests will catch concurrency bugs only once a week or once a month, and we will spend hours and hours chasing seemingly unrelated crashes, like a detective trying to catch a serial killer.

Schema changes should probably be isolated (i.e. external components should either observe both or none), so make this your null hypothesis, and assume it unless proven otherwise.

**Comment by kbr-scylla** [comment]
Why not yield between each base+view pair, but update the pair in isolated way?

**Comment by kbr-scylla** [comment]
Ref https://github.com/scylladb/scylladb/pull/20853

**Comment by ndragazis** [comment]
> Definitely not, this is the wrong approach when talking about concurrency issues. Tests will catch concurrency bugs only once a week or once a month

What's preventing us from having unit tests that trigger such race conditions deterministically?

> Schema changes of view+base should probably be isolated (i.e. external components should either observe both or none), ...
> Why not yield between each base+view pair, but update the pair in isolated way?

Given the changes in #20853 I think that would be redundant. In patch https://github.com/nuivall/scylladb/commit/fe510781c13292f2884328b997d320a423e300c6, I see that @nuivall is changing `diff_table_or_view()` so that it no longer publishes the schemas to the schema registry. This is postponed for the `commit()` phase later on.

@nuivall do you confirm? I could send a PR that makes `diff_table_or_view()` asynchronous and depends on yours.

**Comment by nuivall** [comment]
> @nuivall do you confirm? I could send a PR that makes diff_table_or_view() asynchronous and depends on yours.

Yes, it will be fine. Changes are not committed at that point and schema lock prevents from new changes while yielding.

**Comment by kostja** [comment]
@nuivall is rewriting this piece and his patch should address this.

**Comment by bhalevy** [comment]
> [@nuivall](https://github.com/nuivall) is rewriting this piece and his patch should address this.

ping @nuivall, how are we doing with this?

**Comment by ptrsmrn** [comment]
> > Definitely not, this is the wrong approach when talking about concurrency issues. Tests will catch concurrency bugs only once a week or once a month
> 
> What's preventing us from having unit tests that trigger such race conditions deterministically?
> 
> > Schema changes of view+base should probably be isolated (i.e. external components should either observe both or none), ...
> > Why not yield between each base+view pair, but update the pair in isolated way?
> 
> Given the changes in [#20853](https://github.com/scylladb/scylladb/pull/20853) I think that would be redundant. In patch [nuivall@fe51078](https://github.com/nuivall/scylladb/commit/fe510781c13292f2884328b997d320a423e300c6), I see that [@nuivall](https://github.com/nuivall) is changing `diff_table_or_view()` so that it no longer publishes the schemas to the schema registry. This is postponed for the `commit()` phase later on.

@bhalevy https://github.com/scylladb/scylladb/pull/20853 by @nuivall will most likely resolve the issue.

> 
> [@nuivall](https://github.com/nuivall) do you confirm? I could send a PR that makes `diff_table_or_view()` asynchronous and depends on yours.

**Comment by ptrsmrn** [comment]
@nuivall is rebasing the changes, there'll be an update soon.

**Comment by ptrsmrn** [comment]
The PR had to be dequeued due to an SCT failure, once that's fixed we'll re-merge the PR.

**Comment by mykaul** [comment]
@ptrsmrn - let's follow up on this.

**Comment by ptrsmrn** [comment]
Closing per https://github.com/scylladb/scylladb/issues/21646#issuecomment-2911806808, if the stalls reappear, we'll reopen this issue and investigate.
